python/dolfinx/fem/forms.py

```diff
@@ -20,11 +20,16 @@
 
 if typing.TYPE_CHECKING:
     from dolfinx.fem import function
-    from dolfinx.mesh import Mesh
 
 
 class Form:
-    def __init__(self, form, ufcx_form=None, code: typing.Optional[str] = None):
+    _cpp_object: typing.Tuple[_cpp.fem.Form_complex64, _cpp.fem.Form_complex128,
+                              _cpp.fem.Form_float32, _cpp.fem.Form_float64]
+    _code: typing.Optional[str] = None
```

Comment:
I don't understand some of the typing changes ins this PR. I'll pick a few in the next few comments.

By assigning `_code = None`, isn't this declaring `_code` to be a class attribute, rather than an instance attribute?

Reply:
Yes, until it gets set on an instance, at which point it becomes an attribute of that instance. See the summary at https://stackoverflow.com/a/34126204 of a longer post at https://www.toptal.com/python/python-class-attributes-an-overly-thorough-guide.